sentry-types/src/crontab_validator.rs

```diff
@@ -0,0 +1,79 @@
+use std::collections::HashSet;
+
+#[derive(PartialEq, Eq, Hash)]
+enum CronToken<'a> {
+    Numeric(u64),
+    Alphabetic(&'a str),
+}
+
+const MONTHS: &[&str] = &[
+    "jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec",
+];
+
+const DAYS: &[&str] = &["sun", "mon", "tue", "wed", "thu", "fri", "sat"];
+
+fn value_is_allowed(value: &str, allowed_values: &HashSet<CronToken>) -> bool {
+    match value.parse::<u64>() {
+        Ok(numeric_value) => allowed_values.contains(&CronToken::Numeric(numeric_value)),
+        Err(_) => allowed_values.contains(&CronToken::Alphabetic(&value.to_lowercase())),
+    }
+}
+
+fn validate_range(range: &str, allowed_values: &HashSet<CronToken>) -> bool {
+    range == "*"
+        || range // TODO: Validate that the last range bound is after the previous one.
+            .splitn(2, "-")
+            .all(|bound| value_is_allowed(bound, allowed_values))
+}
+
+/// A valid step is None or Some positive value
+fn validate_step(step: &Option<&str>) -> bool {
+    match *step {
+        Some(value) => match value.parse::<u64>() {
+            Ok(value) => value > 0,
+            Err(_) => false,
+        },
+        None => true,
+    }
+}
+
+fn validate_steprange(steprange: &str, allowed_values: &HashSet<CronToken>) -> bool {
+    let mut steprange_split = steprange.splitn(2, "/");
+    let range = match steprange_split.next() {
+        Some(range) => range,
+        None => {
+            return false;
+        }
+    };
+    let range_is_valid = validate_range(range, allowed_values);
+    let step = steprange_split.next();
+
+    range_is_valid && validate_step(&step)
+}
+
+fn validate_segment(segment: &str, allowed_values: &HashSet<CronToken>) -> bool {
+    segment
+        .split(",")
+        .all(|steprange| validate_steprange(steprange, &allowed_values))
+}
+
+pub fn validate(crontab: &str) -> bool {
+    let allowed_values = vec![
```

Comment:
maybe you can use a static `once_cell` to avoid doing all these allocations every time:
https://docs.rs/once_cell/latest/once_cell/#lazy-initialized-global-data

Reply:
I think caching the allowed_values vector would be unnecessary for my proposed use case; the `sentry-cli` would be calling this function at most once per run in order to validate the crontab passed in by the user. So unless we anticipate another use case where we call `validate` many times per run, I think we can safely keep the function as is.

Reply:
I rewrote this a bit; now the allowed values are defined in a constant